test/jdk/java/awt/color/NonICCFilterTest.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.Color;
+import java.awt.GradientPaint;
+import java.awt.Graphics2D;
+import java.awt.Transparency;
+import java.awt.color.ColorSpace;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorConvertOp;
+import java.awt.image.ComponentColorModel;
+import java.awt.image.DataBuffer;
+import java.awt.image.WritableRaster;
+
+/*
+ * @test
+ * @bug 8316497
+ * @summary Verifies Color filter on Non ICC profile
+ */
+public class NonICCFilterTest {
+
+    private static class TestColorSpace extends ColorSpace {
+
+        private ColorSpace csRGB;
+
+        protected TestColorSpace(boolean bSrc) {
+            super(CS_sRGB, 3);
+            csRGB = ColorSpace.getInstance(bSrc ? ColorSpace.CS_LINEAR_RGB :
+                    ColorSpace.CS_sRGB);
+        }
+
+        public float[] toRGB(float[] colorvalue) {
+            return colorvalue;
```

Comment:
Just for completeness, it is probably better to use csRGB.toRGB and csRGB.fromRGB in these two methods

Reply:
Updated this

Reply:
Hm… If modify the `toRGB` and `fromRGB` methods to use `csRGB` object, then the test does not fail without the fix.

What's more interesting is that the images are equal before `filter` is applied.

Reply:
@aivanov-jdk Thank you for clarifying, I thought he mean to modify `toCIEXYZ` and `fromCIEXYZ`. Yes if we modify `toRGB `and `fromRGB` then it will not pass through fixed area, So we may need to skip this.

Reply:
The difference between srcColorSpace.fromCIEXYZ and dtsColorSpace.fromCIEXYZ can trigger an exception if both have different number of components, for example srcColorSpace = CS_GRAY and srcColorSpace.CS_sRGB =>ArrayIndexOutOfBoundsException will be triggered.

```
    private static TestColorSpace createCS(boolean isSrc) {
        ColorSpace cs = ColorSpace.getInstance(
                isSrc ? ColorSpace.CS_GRAY : ColorSpace.CS_sRGB);
        return new TestColorSpace(cs);
    }

    public static void main(String[] args) {
        BufferedImage src = createTestImage(true);
        BufferedImage dest = createTestImage(false);

        ColorSpace mid = ColorSpace.getInstance(ColorSpace.CS_sRGB);
        ColorConvertOp test = new ColorConvertOp(mid, null);
        test.filter(src, dest);
    }

Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.desktop/java.awt.image.ColorConvertOp.nonICCBIFilter(ColorConvertOp.java:821)
	at java.desktop/java.awt.image.ColorConvertOp.filter(ColorConvertOp.java:275)
	at NonICCFilterTest.main(NonICCFilterTest.java:97)
```

Reply:
and instead of comparing that the images are just different, it is probably better to compare results of conversion via TestColorSpace wrapper vs ColorSpace w/o wrapper.

Reply:
Something like this:
```
/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

import java.awt.Color;
import java.awt.GradientPaint;
import java.awt.Graphics2D;
import java.awt.Transparency;
import java.awt.color.ColorSpace;
import java.awt.image.BufferedImage;
import java.awt.image.ColorConvertOp;
import java.awt.image.ComponentColorModel;
import java.awt.image.DataBuffer;
import java.awt.image.WritableRaster;

/*
 * @test
 * @bug 8316497
 * @summary Verifies Color filter on Non ICC profile
 */
public final class NonICCFilterTest_123 {

    private static final class TestColorSpace extends ColorSpace {

        private final ColorSpace cs;

        TestColorSpace(ColorSpace cs) {
            super(cs.getType(), cs.getNumComponents());
            this.cs = cs;
        }

        public float[] toRGB(float[] colorvalue) {
            return cs.toRGB(colorvalue);
        }

        public float[] fromRGB(float[] rgbvalue) {
            return cs.fromRGB(rgbvalue);
        }

        public float[] toCIEXYZ(float[] colorvalue) {
            return cs.toCIEXYZ(colorvalue);
        }

        public float[] fromCIEXYZ(float[] xyzvalue) {
            return cs.fromCIEXYZ(xyzvalue);
        }
    }

    private static BufferedImage createTestImage(boolean isSrc, boolean plain) {
        ColorSpace cs = createCS(isSrc, plain);
        ComponentColorModel cm = new ComponentColorModel(cs, false, false,
                Transparency.OPAQUE, DataBuffer.TYPE_BYTE);
        WritableRaster raster = cm.createCompatibleWritableRaster(100, 100);
        BufferedImage img = new BufferedImage(cm, raster, false, null);

        Graphics2D g = img.createGraphics();
        GradientPaint gp = new GradientPaint(0, 0, Color.GREEN,
                raster.getWidth(), raster.getHeight(), Color.BLUE);
        g.setPaint(gp);
        g.fillRect(0, 0, raster.getWidth(), raster.getHeight());
        g.dispose();

        return img;
    }

    private static ColorSpace createCS(boolean isSrc, boolean plain) {
        ColorSpace cs = ColorSpace.getInstance(
                isSrc ? ColorSpace.CS_GRAY : ColorSpace.CS_sRGB);
        if (plain) {
            return cs;
        }
        return new TestColorSpace(cs);
    }

    private static boolean compareImages(BufferedImage destTest, BufferedImage destGold) {
        for (int x = 0; x < destTest.getWidth(); x++) {
            for (int y = 0; y < destTest.getHeight(); y++) {
                int rgb1 = destTest.getRGB(x, y);
                int rgb2 = destGold.getRGB(x, y);
                if (rgb1 != rgb2) {
                    System.err.println("rgb1 = " + Integer.toHexString(rgb1));
                    System.err.println("rgb2 = " + Integer.toHexString(rgb2));
                    return true;
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        BufferedImage srcTest = createTestImage(true, false);
        BufferedImage destTest = createTestImage(false, false);

        BufferedImage srcGold = createTestImage(true, true);
        BufferedImage destGold = createTestImage(false, true);

        ColorSpace mid = ColorSpace.getInstance(ColorSpace.CS_sRGB);
        ColorConvertOp test = new ColorConvertOp(mid, null);
        test.filter(srcTest, destTest);

        ColorConvertOp gold = new ColorConvertOp(mid, null);
        gold.filter(srcGold, destGold);

        if (compareImages(destTest, destGold)) {
            throw new RuntimeException("Test failed");
        }
    }
}
```

Note that comparison at the end of the test fails, you can comment it out for now and create a separate bug -> looks like for the case above the results of out conversion is completely unexpected.

Reply:
@mrserb, Thank you for bringing this up. Sorry I am not pretty clear about your suggestion on _Conversion via TestColorSpace wrapper vs ColorSpace w/o wrapper._ . In existing code `compareImages` (pixel value compare) function is local, are you suggesting to move this into wrapper?

Reply:
Your `TestColorSpace` class is a wrapper on top of the actual color space. So you can compare the results of conversion using `TestColorSpace` vs using "actual color space" directly.

Reply:
Thank you for sharing the detailed code, tried the above sample but its like Aleksei mentioned above _always passing irrespective of fix_. The fix area will pass only when non-ICC src and dst. Some of the color related function internally applying icc profile or other propertied and pass through other implementation. Not sure any other way to overcome this.

Reply:
I thought about it… I thought that the filter itself should exercise the code path that's modified but it doesn't seem like it does… I mean creating images with regular ICC profile and using the wrapper `TestColorSpace` in conversion: `new ColorConvertOp(new TestColorSpace(ColorSpace.getInstance(CS_LINEAR_RGB), null);` but it didn't work, the test passes without the fix.

Reply:
The ArrayIndexOutOfBoundsException is the bug we are working on in this PR. The "unexpected behavior" we get when we use the TestColorSpace instead of direct usage of built-in profiles - it looks like the codepath we updated has some other bug that affects the rendering results.

Reply:
>I thought about it… I thought that the filter itself should exercise the code path that's modified but it doesn't seem like it does… I mean creating images with regular ICC profile and using the wrapper TestColorSpace in conversion: new ColorConvertOp(new TestColorSpace(ColorSpace.getInstance(CS_LINEAR_RGB), null); but it didn't work, the test passes without the fix.

It is passed since it blindly checks the difference between src and dst, and since with proper implementation of TestColorSpacewe we always do "some" conversion -> the images are different -> the test passed.

Reply:
> > I thought about it… I thought that the filter itself should exercise the code path that's modified but it doesn't seem like it does… I mean creating images with regular ICC profile and using the wrapper TestColorSpace in conversion: new ColorConvertOp(new TestColorSpace(ColorSpace.getInstance(CS_LINEAR_RGB), null); but it didn't work, the test passes without the fix.
> 
> It is passed since it blindly checks the difference between src and dst, and since with proper implementation of TestColorSpacewe we always do "some" conversion -> the images are different -> the test passed.

Still, the current test catches the typo in the code `srcColorSpace` → `dstColorSpace` that's now replaced. From this point view, the stated bug is fixed.

At the same time, I agree there seem to be more bugs in the area.

Reply:
> Still, the current test catches the typo in the code srcColorSpace → dstColorSpace that's now replaced. From this point view, the stated bug is fixed.

But the implementation of TestColorSpacewe is wrong, and it may not work as expected after any other future updates.

Reply:
There are at least two similar copy/paste typos.
In the next line https://github.com/openjdk/jdk/blob/f73dbb985823d2d59bfcff8ba6951a0d9eb2cefc/src/java.desktop/share/classes/java/awt/image/ColorConvertOp.java#L771 we use the srcNumComp which is the number of components for the incoming source color space, but the loop should be done over iccSrcNumComp which is a number of components of the "intermidiate" source color space. Similar issue is in the next line but for the destination https://github.com/openjdk/jdk/blob/f73dbb985823d2d59bfcff8ba6951a0d9eb2cefc/src/java.desktop/share/classes/java/awt/image/ColorConvertOp.java#L785

It is possible that these compensate/affects each other and produce some non-completly broken result.
The test from the comment above start to pass after these will be fixed. 

Since we found a few bugs in this code path, it will be useful to check the code path in the nonICCBIFilter where the CSList is not null(the intermidiate transform is wrapper as well, as of now our test uses the built-in ColorSpace.getInstance(CS_sRGB))

Reply:
> There are at least two similar copy/paste typos. In the next line
> 
> https://github.com/openjdk/jdk/blob/f73dbb985823d2d59bfcff8ba6951a0d9eb2cefc/src/java.desktop/share/classes/java/awt/image/ColorConvertOp.java#L771
> we use the srcNumComp which is the number of components for the incoming source color space, but the loop should be done over iccSrcNumComp which is a number of components of the "intermidiate" source color space. Similar issue is in the next line but for the destination
> 
> https://github.com/openjdk/jdk/blob/f73dbb985823d2d59bfcff8ba6951a0d9eb2cefc/src/java.desktop/share/classes/java/awt/image/ColorConvertOp.java#L785
> 
> It is possible that these compensate/affects each other and produce some non-completly broken result. The test from the comment above start to pass after these will be fixed.

Indeed, these should be `iccSrcNumComp` and `iccDstNumComp` according to number of components in the allocated arrays. Modifying these lines make the test pass.

> Since we found a few bugs in this code path, it will be useful to check the code path in the nonICCBIFilter where the CSList is not null(the intermidiate transform is wrapper as well, as of now our test uses the built-in ColorSpace.getInstance(CS_sRGB))

Agree, another (unit) test is required which exercises other code paths.

Reply:
These typos also fixed